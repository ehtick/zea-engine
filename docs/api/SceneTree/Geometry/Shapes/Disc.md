<a name="Disc"></a>

### Disc 
A class for generating a disc geometry.

```
const disc = new Disc(2.0, 22)
```

**Parameters**
* **radius(`NumberParameter`):** Specifies the radius of the disc.
* **sides(`NumberParameter`):** Specifies the resolution, or the disc subdivisions around `Z` axis.


**Extends**: <code>Mesh</code>  

* [Disc ⇐ <code>Mesh</code>](#Disc)
    * [new Disc([radius], [sides])](#new-Disc)
    * [radius ⇒ <code>number</code>](#radius)
    * [radius](#radius)
    * [sides](#sides)
    * [toJSON() ⇒ <code>object</code>](#toJSON)

<a name="new_Disc_new"></a>

### new Disc
Creates an instance of Disc.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [radius] | <code>number</code> | <code>0.5</code> | The radius of the disc. |
| [sides] | <code>number</code> | <code>32</code> | The number of sides. |

<a name="Disc+radius"></a>

### radius 
Returns the value of the `radius` parameter.


**Returns**: <code>number</code> - - Returns the radius.  
<a name="Disc+radius"></a>

### radius
Sets the value of the `radius` parameter.



| Param | Type | Description |
| --- | --- | --- |
| val | <code>number</code> | The radius value. |

<a name="Disc+sides"></a>

### sides
Sets the value of the `sides` parameter.



| Param | Type | Description |
| --- | --- | --- |
| val | <code>number</code> | The number of sides. |

<a name="Disc+toJSON"></a>

### toJSON
The toJSON method encodes this type as a json object for persistence.


**Returns**: <code>object</code> - - Returns the json object.  
