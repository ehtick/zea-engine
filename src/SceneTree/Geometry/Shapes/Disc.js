import { Vec2 } from '../../../Math/Vec2'
import { Vec3 } from '../../../Math/Vec3'
import { NumberParameter } from '../../Parameters/index'
import Registry from '../../../Registry'
import ProceduralMesh from './ProceduralMesh'

/**
 * A class for generating a disc geometry.
 *
 * ```
 * const disc = new Disc(2.0, 22)
 * ```
 *
 * **Parameters**
 * * **Radius(`NumberParameter`):** Specifies the radius of the disc.
 * * **Sides(`NumberParameter`):** Specifies the resolution, or the disc subdivisions around `Z` axis.
 *
 * @extends {ProceduralMesh}
 */
class Disc extends ProceduralMesh {
  /**
   * Creates an instance of Disc.
   *
   * @param {number} [radius=0.5] - The radius of the disc.
   * @param {number} [sides=32] - The number of sides.
   */
  constructor(radius = 0.5, sides = 32) {
    super()

    if (isNaN(radius) || isNaN(sides)) throw new Error('Invalid geom args')

    this.__radiusParam = this.addParameter(new NumberParameter('Radius', radius))
    this.__sidesParam = this.addParameter(new NumberParameter('Sides', sides >= 3 ? sides : 3, [3, 200], 1))

    this.addVertexAttribute('texCoords', Vec2)
    this.addVertexAttribute('normals', Vec3)
    this.__rebuild()
  }

  /**
   * Returns the value of the `radius` parameter.
   *
   * @return {number} - Returns the radius.
   */
  get radius() {
    return this.__radius
  }

  /**
   * Sets the value of the `radius` parameter.
   *
   * @param {number} val - The radius value.
   */
  set radius(val) {
    this.__radius = val
    this.__resize()
  }

  /**
   * Sets the value of the `sides` parameter.
   * @param {number} val - The number of sides.
   */
  set sides(val) {
    this.__sides = val >= 3 ? val : 3
    this.__rebuild()
  }

  /**
   * The __rebuild method.
   * @private
   */
  __rebuild() {
    const nbSides = this.__sidesParam.getValue()

    this.setNumVertices(nbSides + 1)
    this.setFaceCounts([nbSides])

    // ////////////////////////////
    // Set Vertex Positions
    const positions = this.getVertexAttribute('positions')
    positions.getValueRef(0).set(0.0, 0.0, 0.0)

    // ////////////////////////////
    // Build the topology
    for (let j = 0; j < nbSides; j++) {
      const v1 = (j % nbSides) + 1
      const v2 = ((j + 1) % nbSides) + 1
      this.setFaceVertexIndices(j, [0, v1, v2])
    }

    // ////////////////////////////
    // setNormals
    const normals = this.getVertexAttribute('normals')
    // Now set the attrbute values
    const normal = new Vec3(0, 0, 1)
    normals.setValue(0, normal)
    for (let i = 0; i < nbSides; i++) {
      normals.setValue(i + 1, normal)
    }

    // ////////////////////////////
    // setUVs
    const texCoords = this.getVertexAttribute('texCoords')
    texCoords.getValueRef(0).set(0.5, 0.5)
    for (let i = 0; i < nbSides; i++) {
      const phi = (i / nbSides) * 2.0 * Math.PI
      texCoords.getValueRef(i + 1).set(Math.sin(phi) * 0.5 + 0.5, Math.cos(phi) * 0.5 + 0.5)
    }

    this.setBoundingBoxDirty()
    this.__resize()
  }

  /**
   * The __resize method.
   * @private
   */
  __resize() {
    const nbSides = this.__sidesParam.getValue()
    const radius = this.__radiusParam.getValue()
    const positions = this.getVertexAttribute('positions')
    for (let i = 0; i < nbSides; i++) {
      const phi = (i / nbSides) * 2.0 * Math.PI
      positions.getValueRef(i + 1).set(Math.sin(phi) * radius, Math.cos(phi) * radius, 0.0)
    }
    this.setBoundingBoxDirty()
    this.emit('geomDataChanged', {})
  }
}

Registry.register('Disc', Disc)

export { Disc }
