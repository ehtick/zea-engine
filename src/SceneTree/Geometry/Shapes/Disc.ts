import { Vec2 } from '../../../Math/Vec2'
import { Vec3 } from '../../../Math/Vec3'
import { NumberParameter } from '../../Parameters/index'
import { Registry } from '../../../Registry'
import { ProceduralMesh } from './ProceduralMesh'
import { Vec2Attribute } from '../Vec2Attribute'
import { Vec3Attribute } from '../Vec3Attribute'

/**
 * A class for generating a disc geometry.
 *
 * ```
 * const disc = new Disc(2.0, 22)
 * ```
 *
 * **Parameters**
 * * **Radius(`NumberParameter`):** Specifies the radius of the disc.
 * * **Sides(`NumberParameter`):** Specifies the resolution, or the disc subdivisions around `Z` axis.
 *
 * @extends {ProceduralMesh}
 */
class Disc extends ProceduralMesh {
  protected __radiusParam: NumberParameter
  protected __sidesParam: NumberParameter
  topologyParams: string[]

  /**
   * Creates an instance of Disc.
   *
   * @param {number} [radius=0.5] - The radius of the disc.
   * @param {number} [sides=32] - The number of sides.
   */
  constructor(radius = 0.5, sides = 32) {
    super()
    this.topologyParams = []
    if (isNaN(radius) || isNaN(sides)) throw new Error('Invalid geom args')

    this.__radiusParam = this.addParameter(new NumberParameter('Radius', radius)) as NumberParameter
    this.__sidesParam = this.addParameter(
      new NumberParameter('Sides', sides >= 3 ? sides : 3, [3, 200], 1)
    ) as NumberParameter

    this.addVertexAttribute('texCoords', Vec2)
    this.addVertexAttribute('normals', Vec3)

    this.topologyParams.push('Sides')
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const nbSides = this.__sidesParam.getValue() || 32

    this.setNumVertices(nbSides + 1)
    this.setFaceCounts([nbSides])

    // ////////////////////////////
    // Set Vertex Positions
    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    if (positions) positions.getValueRef(0).set(0.0, 0.0, 0.0)

    // ////////////////////////////
    // Build the topology
    for (let j = 0; j < nbSides; j++) {
      const v1 = (j % nbSides) + 1
      const v2 = ((j + 1) % nbSides) + 1
      this.setFaceVertexIndices(j, [0, v1, v2])
    }

    // ////////////////////////////
    // setNormals
    const normals = <Vec3Attribute>this.getVertexAttribute('normals')
    if (normals) {
      // Now set the attrbute values
      const normal = new Vec3(0, 0, 1)
      normals.setValue(0, normal)
      for (let i = 0; i < nbSides; i++) {
        normals.setValue(i + 1, normal)
      }
    }

    // ////////////////////////////
    // setUVs
    const texCoords = <Vec2Attribute>this.getVertexAttribute('texCoords')
    if (texCoords) {
      texCoords.getValueRef(0).set(0.5, 0.5)
      for (let i = 0; i < nbSides; i++) {
        const phi = (i / nbSides) * 2.0 * Math.PI
        texCoords.getValueRef(i + 1).set(Math.sin(phi) * 0.5 + 0.5, Math.cos(phi) * 0.5 + 0.5)
      }
    }

    this.resize()
  }

  /**
   * The resize method.
   * @private
   */
  resize(): void {
    const nbSides = this.__sidesParam.getValue() || 32
    const radius = this.__radiusParam.getValue() || 0.5
    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    if (positions) {
      for (let i = 0; i < nbSides; i++) {
        const phi = (i / nbSides) * 2.0 * Math.PI
        positions.getValueRef(i + 1).set(Math.sin(phi) * radius, Math.cos(phi) * radius, 0.0)
      }
    }
  }
}

Registry.register('Disc', Disc)

export { Disc }
