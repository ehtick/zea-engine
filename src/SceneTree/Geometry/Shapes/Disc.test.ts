import { Disc } from './Disc'

describe('Disc', () => {
  it('tests default parameters', () => {
    const disc = new Disc()

    expect(disc.radiusParam.value).toBe(0.5)
    expect(disc.sidesParam.value).toBe(32)
  })

  it('updates parameters', () => {
    const disc = new Disc()
    disc.radiusParam.value = 4
    disc.sidesParam.value = 16

    expect(disc.radiusParam.value).toBe(4)
    expect(disc.sidesParam.value).toBe(16)
  })

  it('saves to JSON (serialization).', () => {
    const disc = new Disc(3, 16)
    const outputJSON = disc.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const disc = new Disc()
    const jsonStr = {
      params: {
        Radius: {
          value: 3
        },
        Sides: {
          value: 16,
          range: [3, 200],
          step: 1
        }
      },
      type: 'Disc',
      vertexAttributes: {}
    }
    disc.fromJSON(jsonStr)

    const newDisc = new Disc(3, 16)
    expect(disc.toJSON()).toEqual(newDisc.toJSON())
  })
})
