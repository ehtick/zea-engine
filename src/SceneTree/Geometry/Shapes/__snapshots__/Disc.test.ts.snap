// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Disc saves to JSON (serialization). 1`] = `
Object {
  "params": Object {
    "Radius": Object {
      "value": 3,
    },
    "Sides": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 16,
    },
  },
  "type": "Disc",
  "vertexAttributes": Object {},
}
`;
